feat!: Partition based on non-numeric and multiple keys

sundar-mudupalli-work opened the pull request:
This PR includes code, testing and documentation changes to partition tables based on non-numeric and multiple keys. The partitioned tables can be used to perform `validate row` and possibly `grouped-columns` validations without running into `MemoryError` issues.

One system test was added to the postgres test set to partition a small database with multiple numeric and string keys. The results are checked against the expected value. One unit test was deleted that was written for partitioning a numeric key in pandas - the updated approach with ntile does not work on pandas tables. 

This code works correctly with Oracle and Postgres databases - has not been tested with other databases.

BREAKING CHANGE:  generate-table-partitions no longer supports a partition-key parameter, since the primary key(s) are used to partition the table.

sundar-mudupalli-work committed:
Changes to support non numeric partitions

sundar-mudupalli-work committed:
This commit implements partitioning using NTILE database function. The PRD
for this feature is at https://docs.google.com/document/d/14ljODhkXFRNAbU5JchwlY_AOdx3bz-qnbH1BcnGM0ss/edit?resourcekey=0-lFpqOEk04nMAKN-HkJdERg

Key Benefits:
1. Database NTILE function is being used - which splits the rows into the specified set of partitions (batches) - this is done exactly for most
   relational databases. BigQuery has an approximate NTILE function - this code has not been tested with BigQuery yet.
2. It supports any column type that can be collated (integer, text, boolean, I believe)
3. You can use multiple partition keys (the generated where clauses look complex, and should work).

Drawbacks:
1. NTILE function is efficient, still this likely results in something similar to a full table scan
2. Not all databases support NTILE

sundar-mudupalli-work committed:
Updated to remove --partition-key as an optional argument, since primary keys are used and partition-key is ignored
Updated partition logic to use the first element of each partition - please see partitio logic PRD.

sundar-mudupalli-work committed:
Merge branch 'develop' into issue688-non-numeric-partitions

Getting branch up to date with develop

sundar-mudupalli-work committed:
Updated files to run nox tests - most tests pass - except one requiring windowing support in Pandas
This is not implemented.

sundar-mudupalli-work committed:
Design doc for paritition table in markdown format.

sundar-mudupalli-work committed:
Updated doc regarding changes to generate-table-partitions

sundar-mudupalli-work committed:
Updated code to point to internal design doc. Cleaned up internal design doc

svetakvsundhar reviewed (commented):
Thanks. I'm not completely familiar with how this tool works, so I will defer to someone with more knowledge of the tool for a review.

renzokuken commented:
/gcbrun

renzokuken committed:
resolving merge conflict

sundar-mudupalli-work committed:
Added definition of a table used to generate table partitions.
If this file is run on a clean cloudsql instannce, the table required for generating table partitions will be created.
However, I don't see the table definitions of the other tables used in tests in this file.
Perhaps the assumption is that database servers will be backed up and providing sql scripts to generate tables is not required.

sundar-mudupalli-work committed:
Added function to test generate table partitions.
Queries a table and check the generated partition clauses against what is expected.

sundar-mudupalli-work committed:
test was failing because the test_get_pandas_data_client was expecting an argument (or a fixture) that was never used
Deleted the argument/fixture and test seems to be running correctly.

sundar-mudupalli-work committed:
Updated unit tests - now tests the parameters provided in the unit test
Actual checking of the partitioning logic is now checked in the system test as a real database is required
for that functionality.

sundar-mudupalli-work committed:
Merge branch 'issue688-non-numeric-partitions' of https://github.com/GoogleCloudPlatform/professional-services-data-validator into issue688-non-numeric-partitions

sundar-mudupalli-work commented:
/gcbrun

renzokuken commented on data_validation/query_builder/partition_row_builder.py:
```diff
@@ -20,7 +20,7 @@
     def __init__(
         self,
-        partition_key: str,
+        primary_keys: [str],
```
@sundar-mudupalli-work, what is the rationale for changing the partition key argument to primary key? I'm not opposed just worried that primary key may be referenced in other paths of the tool

sundar-mudupalli-work replied:
We use the primary-key(s) to match corresponding rows between the source and target. When we partition the table, we only compare one partition in the source with the equivalent (corresponding) partition in the target. So ensuring the same set of primary keys fall into the same partition in both the source and target is necessary and sufficient. If the partition-key uniquely identifies the row, then the partition key can be used instead of the primary key. I am going to update the document on this point - the primary key is only used for matching rows and does not need to be a `primary` key in the table schema. When we say primary key, it can be any column (or set of columns) that uniquely identifies the row.

sundar-mudupalli-work replied:
change that to sequence of columns - the order is important.

renzokuken commented on tests/unit/test_clients.py:
```diff
@@ -85,7 +85,7 @@ def test_get_oracle_data_client():
 
 
-def test_get_pandas_data_client(fs):
+def test_get_pandas_data_client():
```
is this just cleanup or was the test failing?

sundar-mudupalli-work replied:
I saw an error one time saying fs is not a valid fixture. I can no longer reproduce it. Therefore consider it a cleanup.

renzokuken reviewed (commented):
Just a couple of questions in-line. Otherwise LGTM

renzokuken reviewed (approved):
Approved peinding primary key definition is updated. LGTM

sundar-mudupalli-work committed:
Updated README.md to clarify what primary keys refers to

renzokuken commented:
/gcbrun

nehanene15 commented:
Looks great Sundar! One comment - can we also document that we only support Oracle and Postgres for this feature? Also, did you want to refer to the internal partition PRD you created in the main README?
